src/lib/listeners/EngineProgressListener.ts

```diff
@@ -0,0 +1,105 @@
+import {
+	CodeAnalyzer,
+	EngineProgressEvent,
+	EngineResultsEvent,
+	EventType,
+	RuleSelection
+} from '@salesforce/code-analyzer-core';
+import {Display} from '../Display';
+import {BundleName, getMessage} from '../messages';
+
+export interface EngineProgressListener {
+	listen(codeAnalyzer: CodeAnalyzer, ruleSelection: RuleSelection): void;
+	stopListening(): void;
+}
+
+export class SpinnerProgressListener implements EngineProgressListener {
+	private readonly display: Display;
+	private readonly tickTime: number;
+	/**
+	 * Mapping from each engine's name to a number indicating its completion percentage.
+	 * @private
+	 */
+	private readonly progressMap: Map<string, number>;
+	private startTime: number;
+	private intervalId: NodeJS.Timeout;
+
+	/**
+	 *
+	 * @param display The display with which to show output to the user.
+	 * @param tickTime A number of milliseconds to wait between automatic ticks. Defaults to 1000. Negative values disable ticking.
+	 */
+	public constructor(display: Display, tickTime: number = 1000) {
+		this.display = display;
+		this.tickTime = tickTime;
+		this.progressMap = new Map();
+	}
+
+	public listen(codeAnalyzer: CodeAnalyzer, ruleSelection: RuleSelection): void {
+		this.startTime = Date.now();
+		this.startSpinning(ruleSelection);
+		codeAnalyzer.onEvent(EventType.EngineProgressEvent, e => this.updateCorrespondingEngine(e as EngineProgressEvent));
+		codeAnalyzer.onEvent(EventType.EngineResultsEvent, e => this.completeCorrespondingEngine(e as EngineResultsEvent));
```

Comment:
Instead of casting, you have some choices, you either do:
* codeAnalyzer.onEvent<EngineProgressEvent>(..., e => ...)
* or codeAnalyzer.onEvent(..., (e: EngineProgressEvent) => ...)

If you go with the first approach you might even be able to just have it say:

codeAnalyzer.onEvent<EngineProgressEvent>(EventType.EngineProgressEvent, this.updateCorrespondingEngine);